### ISSUE TITLE
bad complexity of Array functions

### ISSUE THREAD

**Issue by kuribas** [purescript/purescript-arrays#13]
Because Array.slice has complexity 0(n), many of the functions in Data.Array have a worse complexity than they should have.
head is O(n) instead of O(1)
last is O(n^2) where it could be O(1)
tail is O(n) (due to copying instead of sharing)

I find that pattern matching on Arrays using cons syntax is generally a bad idea.  It almost always leads to worse performance than expected.  Functional programmers are used to pattern matching being free (O(1)).  By having
cons patterns which are O(n), it may confuse programmers and give unexpected performance penalties.
The examples shown on the patterns documentation (sum and addPairs) are also O(n^2).

**Comment by jacereda** [comment]
@kuribas do you mean tail can be implemented better than O(n)? How?

**Comment by kuribas** [comment]
@jacereda if you change the representation from an Array as an range and a JS array, like how vectors are shared in the vector package in haskell, but paf31 said that would break the FFI, so basicly no.

**Comment by paf31** [comment]
It wouldn't break the FFI, but it is a different type, so it needs a different foreign type. It would make a nice library though.

**Comment by jacereda** [comment]
Something like this?

https://github.com/jacereda/purescript-slices

**Comment by paf31** [comment]
I like it :)

**Comment by hdgarrood** [comment]
I just posted on the mailing list about this issue - it might be of interest. https://groups.google.com/forum/#!topic/purescript/uF-SqphMeEg

TLDR - should we be encouraging use of native arrays at all (except for interop), when we can use other data structures which are better suited to functional programs, like Haskell's Data.Sequence?

**Comment by hdgarrood** [comment]
Now that #28 has been merged, and given that deprecating cons syntax is planned on an issue in the purescript repo (purescript/purescript#937), I think this should be closed; I don't think there's anything else that can be done in this library to mitigate this issue.

**Comment by paf31** [comment]
Agreed. We have better data structures for different classes of problems. The issue is just getting people to use them, I suppose.

**Comment by alexchandel** [comment]
It's too bad the purescript optimizer can't combine/optimize the array functions & pattern matching. Isn't `head` just `[0]` in JS?

**Comment by paf31** [comment]
Well, when I finally implement rewrite rules, maybe that will be possible.

**Comment by alexchandel** [comment]
@paf31 wait, that doesn't happen automatically? I guess I assumed that everything except `Eff`s could be subject to rewriting, inlining, or outright deletion by the optimizer.

**Comment by paf31** [comment]
There are a handful of baked-in optimizations, it's not a general-purpose optimizer.

**Comment by garyb** [comment]
Closing this as it's semi-addressed now. Cons patterns for arrays have been removed in the compiler and some work has been done to optimise various array functions.
